Add support for int[]? arguments in native_functions.yaml

dreiss opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #37177 Update interpolate to use new upsample overloads
* #37176 Add interpolate-style overloads to aten::upsample* ops
* #37175 Add support for float[]? arguments in native_functions.yaml
* **#37174 Add support for int[]? arguments in native_functions.yaml**
* #37173 In interpolate, inline the call to _interp_output_size
* #37172 In interpolate, move exceptional cases to the bottom
* #37171 In interpolate, use if instead of elif





Differential Revision: [D21210002](https://our.internmc.facebook.com/intern/diff/D21210002/)

Differential Revision: [D21210002](https://our.internmc.facebook.com/intern/diff/D21210002)

dreiss committed:
Add support for int[]? arguments in native_functions.yaml



Differential Revision: [D21210002](https://our.internmc.facebook.com/intern/diff/D21210002/)

[ghstack-poisoned]

dreiss commented:
Is there a good way to write unit tests for this?

gchanan commented on torch/csrc/utils/python_arg_parser.h:
```diff
@@ -153,6 +153,7 @@ struct PythonArgs {
   template<int N>
   inline std::array<at::Tensor, N> tensorlist_n(int i);
   inline std::vector<int64_t> intlist(int i);
+  inline c10::OptionalArray<int64_t> intlistOptional(int i);
```
this looks fine, although none of these things use c10/IValue types so I'm not sure why we wold change that here.

dreiss replied:
`OptionalArray` isn't really an IValue type.  It's just a vector wrapper that allows implicit conversion to ArrayRef.  I could move it to ATen if you want.

gchanan commented on tools/autograd/templates/Functions.h:
```diff
@@ -17,6 +17,7 @@ namespace torch { namespace autograd { namespace generated {
 using at::Scalar;
 using at::Tensor;
 using at::IntArrayRef;
+using at::ArrayRef;
```
is this necessary?  I don't see any generated code that needs this.

dreiss replied:
Oh, this might have only been needed for the next diff in the stack, which uses `ArrayRef<double>`

gchanan commented on tools/autograd/templates/python_nn_functions.cpp:
```diff
@@ -15,6 +15,7 @@ using at::Scalar;
 using at::MemoryFormat;
 using at::Generator;
 using at::IntArrayRef;
+using at::ArrayRef;
```
same here.

gchanan reviewed (approved):
eager parts look good to me.  I'd have @eellison or @smessmer look at the c10/JIT parts.

gchanan commented:
also, I think it's a good idea to make sure it works end-to-end by just declaring a fake underscore function, in native_functions "_test_intlist_optional" that just returns the optional intlist or something similar that you check works in (at least) eager.

gchanan commented:
(sorry, clicked close and comment instead of comment).

dreiss committed:
Update on "Add support for int[]? arguments in native_functions.yaml"






Differential Revision: [D21210002](https://our.internmc.facebook.com/intern/diff/D21210002/)

Differential Revision: [D21210002](https://our.internmc.facebook.com/intern/diff/D21210002)

[ghstack-poisoned]

dreiss committed:
Update on "Add support for int[]? arguments in native_functions.yaml"






Differential Revision: [D21210002](https://our.internmc.facebook.com/intern/diff/D21210002/)

Differential Revision: [D21210002](https://our.internmc.facebook.com/intern/diff/D21210002)

[ghstack-poisoned]

dreiss committed:
Update on "Add support for int[]? arguments in native_functions.yaml"






Differential Revision: [D21210002](https://our.internmc.facebook.com/intern/diff/D21210002/)

Differential Revision: [D21210002](https://our.internmc.facebook.com/intern/diff/D21210002)

[ghstack-poisoned]

dreiss committed:
Update on "Add support for int[]? arguments in native_functions.yaml"






Differential Revision: [D21210002](https://our.internmc.facebook.com/intern/diff/D21210002/)

Differential Revision: [D21210002](https://our.internmc.facebook.com/intern/diff/D21210002)

[ghstack-poisoned]

dreiss committed:
Update on "Add support for int[]? arguments in native_functions.yaml"






Differential Revision: [D21210002](https://our.internmc.facebook.com/intern/diff/D21210002/)

Differential Revision: [D21210002](https://our.internmc.facebook.com/intern/diff/D21210002)

[ghstack-poisoned]
